### PR TITLE
Refactor widgets to handle dynamic config in input

### LINKED ISSUE #8532: No dynamic widgets in table nodes.
image

Both in `aggregate` and `order_by` nodes, the column/aggregates list has 
* default argument when adding new item
* drop-downs inside list
in GUI1, but not in GUI2

Simpler case like `Table.get` also does not display drop down with list of columns.

### PR DESCRIPTION
**Pull request by farmaazon**
closes https://github.com/enso-org/enso/issues/8532

### Pull Request Description

* Merged ArgumentAst and ArgumentPlaceholder into single class.
* Created `AnyWidget` input being a "general use" widget input. Most wigets try to match with it; the `Argument` input is now solely for WidgetArgument(Name) or those handling arguments in a specific way (like selector which want to show on arg name click).
* dynamic config is now part of widget input, and is properly propagated through vector editor/function widgets.

### Important Notes

The widgets still does not work perfectly:
* The chosen options often don't have argument placeholders - that's because we don't display them for constructors. Needs to be added on our side, or engine should send us methodCall info for constructors.
* There are issues with engine's messages sent to us. This makes widgets does not set up (so there is no drop-down, or vector adds `_` instead of default). I'm investigating them and going to fill issues.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: app/gui2/src/components/GraphEditor/widgets/WidgetFunction.vue [hunk 1]
```diff
@@ -53,19 +68,26 @@ const escapeString = (str: string): string => {
...
+const selfArgumentAstId = computed<Opt<ExprId>>(() => {
+  const analyzed = interpretCall(props.input.ast, true)
   if (analyzed.kind === 'infix') {
     return analyzed.lhs?.exprId
   } else {
-    return analyzed.args[0]?.argument.exprId
+    const knownArguments = methodCallInfo.value?.suggestion?.arguments
+    const selfArgument =
+      knownArguments?.[0]?.name === 'self'
+        ? getAccessOprSubject(analyzed.func)
```
Comment (Frizi) [inline app/gui2/src/components/GraphEditor/widgets/WidgetFunction.vue:78]: There might be a `self` argument applied as named.

File: app/gui2/src/components/GraphEditor/widgets/WidgetFunction.vue [hunk 2]
```diff
@@ -83,14 +105,16 @@ const visualizationConfig = computed<Opt<NodeVisualizationConfiguration>>(() =>
     positionalArgumentsExpressions: [`.${name}`, makeArgsList(args)],
   }
 })
+watch(visualizationConfig, console.log)
```
Comment (Frizi) [inline app/gui2/src/components/GraphEditor/widgets/WidgetFunction.vue]: log

File: app/gui2/src/components/GraphEditor/widgets/WidgetFunction.vue [hunk 3]
```diff
@@ -199,10 +223,22 @@ function handleArgUpdate(value: unknown, origin: PortId): boolean {
 }
 </script>
 <script lang="ts">
-export const widgetDefinition = defineWidget([Ast.App, Ast.Ident, Ast.OprApp], {
+function isFunctionCall(
+  input: WidgetInput,
+): input is AnyWidget & { ast: Ast.App | Ast.Ident | Ast.OprApp } {
+  if (input instanceof AnyWidget)
+    return (
```
Comment (Frizi) [inline app/gui2/src/components/GraphEditor/widgets/WidgetFunction.vue]: simplify, potentially create a `withAst` static method on `AnyWidget` to allow creating matchers for specific AST types.

File: app/gui2/src/components/GraphEditor/widgets/WidgetNumber.vue [hunk 4]
```diff
@@ -1,30 +1,39 @@
...
-import { Score, defineWidget, widgetProps } from '@/providers/widgetRegistry'
+import { AnyWidget, Score, defineWidget, widgetProps } from '@/providers/widgetRegistry'
 import { Ast } from '@/util/ast'
 import { computed } from 'vue'
 
 const props = defineProps(widgetProps(widgetDefinition))
 const value = computed({
   get() {
-    return parseFloat(props.input.code())
+    return parseFloat(props.input.ast?.code() ?? '')
```
Comment (Frizi) [inline app/gui2/src/components/GraphEditor/widgets/WidgetNumber.vue]: That expression will return `NaN` when no AST is present.

File: app/gui2/src/providers/widgetRegistry.ts [hunk 5]
```diff
@@ -1,11 +1,60 @@
...
+import type { SuggestionEntryArgument } from '@/stores/suggestionDatabase/entry'
+import type { Ast } from '@/util/ast'
 import { computed, shallowReactive, type Component, type PropType } from 'vue'
 
 export type WidgetComponent<T extends WidgetInput> = Component<WidgetProps<T>>
 
+/**
+ * A WidgetInput variant meant to match wide range of "general" widgets.
+ *
+ * Any widget which want's to work in different contexts (inside function calls, constructors,
```
Comment (vitvakatu) [inline app/gui2/src/providers/widgetRegistry.ts]: nit: 
```suggestion
 * Any widget which wants to work in different contexts (inside function calls, constructors,
```

File: app/gui2/src/components/GraphEditor/widgets/WidgetFunction.vue (hunk 2 above)
Comment (vitvakatu) [inline app/gui2/src/components/GraphEditor/widgets/WidgetFunction.vue]: nit: debug logging

File: app/gui2/src/components/GraphEditor/widgets/WidgetFunction.vue [hunk 6]
```diff
@@ -53,19 +68,26 @@ const escapeString = (str: string): string => {
...
+    return selfArgument?.exprId
   }
 })
 
 const visualizationConfig = computed<Opt<NodeVisualizationConfiguration>>(() => {
-  const tree = props.input
-  const expressionId = selfArgumentExprId.value
-  const astId = tree.exprId
+  // If we inherit dynamic config, there is no point in attaching visualization.
+  if (props.input.dynamicConfig) return null
```
Comment (vitvakatu) [inline app/gui2/src/components/GraphEditor/widgets/WidgetFunction.vue:86]: Is it correct? If we have a function call in one of the arguments of another function call, we will receive a dynamic config for the outer call only.

Comment (Frizi) [reply app/gui2/src/components/GraphEditor/widgets/WidgetFunction.vue:86]: At this point, the `dynamicConfig` property is scoped to this particular argument value.

**Review by Frizi — APPROVED** [review]

**Review by Frizi — APPROVED** [review]

File: app/gui2/src/components/GraphEditor/widgets/WidgetFunction.vue (hunk 6 above)
Comment (farmaazon) [reply app/gui2/src/components/GraphEditor/widgets/WidgetFunction.vue:86]: The dynamic config for outer call could have quite a detailed config for its argument - for example, the Single_Choice specifies config for parameters of chosen call. I didn't want to make some complex merging system yet. Let's see how often we will be affected.

**Pull request merged**